index.html

```diff
@@ -3549,28 +3548,37 @@ <h3><dfn>Writing or pushing content</dfn></h3>
         the <a>activated reader objects</a>, run the following sub-steps:
         <ol>
           <li>
-            If |reader|.<a>[[\Id]]</a> is [= dictionary member/present =]
-            and it is not equal to any |record|'s
-            <a href="#dom-ndefrecord-id">id</a> where |record|
-            is an element of |message|, [= iteration/continue =].
-          </li>
-          <li>
-            If |reader|.<a>[[\RecordType]]</a> is [= dictionary member/present =]
-            and it is not equal to any |record|'s
-            <a href="#dom-ndefrecord-recordtype">recordType</a> where |record|
-            is an element of |message|, [= iteration/continue =].
-          </li>
-          <li>
-            If |reader|.<a>[[\MediaType]]</a> is not `""` and
-            it is not equal to any |record|'s
-            <a href="#dom-ndefrecord-mediatype">mediaType</a> where |record| is
-            an element of |message|, [= iteration/continue =].
-          </li>
-          <li>
-            <a>Fire an event</a> named "`reading`" at |reader|
-            using <a>NDEFReadingEvent</a> with its <a>serialNumber</a> attribute
-            initialized to |serialNumber| and <a>message</a> attribute
-            initialized to |message|.
+            For each |record:NDEFRecord| in |message:NDEFMessage|,
+            <ol>
+              <li>
+                Let |matched| be `undefined`.
+              </li>
+              <li>
+                If |reader|.<a>[[\Id]]</a> is not `""` and is not equal to
+                |record|'s <a href="#dom-ndefrecord-id">id</a>, set |matched|
+                to `false`. If it is equal, set |matched| to `true`.
```

Comment:
If we found `Id` does not match, we can just skip comparing `RecordType` and `MediaType` and continue with the next record. Like:
For each reader,
  For each record in the message,
    1. If `Id` does not match, continue with the next record;
    2. If `RecordType` does not match, continue with the next record;
    3. If `MediaType` does not match, continue with the next record;
    4. Found a matched record, means the message matches, fire the event to the current reader and continue with the next reader.

Reply:
That was the old algorithm. 
This new one looks for any record in a message that satisfies _all_ defined elements of the filter.

Reply:
But you are right that in this algo, if a filter field is defined but it's not a match, we should continue, not just set `matched` to `false`. I will correct that.